File: src/main/java/com/tkr/ingress/IngressDispatch.java
package com.tkr.ingress;

import com.tkr.nativelink.NativeBridge;
import com.tkr.types.WireTypes;
import com.tkr.types.WireTypes.*;
import com.tkr.util.BoundsUtil;
import com.tkr.util.DigestUtil;
import com.tkr.wire.BatchWireCodec;
import com.tkr.wire.SessionWireCodec;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** Dispatches classified ingress streams into envelope channel views. */
public final class IngressDispatch {

    public static final int ENVELOPE_HEADER_SIZE = 24;
    public static final int ENVELOPE_CHANNEL_SIZE = 20;

    public static final class IngressDispatchConfig {
        public boolean sealOnSweep = true;
        public boolean queueDeferred = true;
        public boolean validateDigests = true;
        public int maxChannels = WireTypes.MAX_ENVELOPE_CHANNELS;
    }

    public static final class IngressStreamResult {
        public Status status = Status.OK;
        public List<IngressEnvelope> completedEnvelopes = new ArrayList<>();
        public int framesParsed;
    }

    private final IngressDispatchConfig config;
    private final IngressClassifier classifier;
    private int nextIngressSeq = 1;

    public IngressDispatch() { this(new IngressDispatchConfig()); }
    public IngressDispatch(IngressDispatchConfig config) {
        this.config = config != null ? config : new IngressDispatchConfig();
        this.classifier = new IngressClassifier();
    }

    public IngressStreamResult processIngressStream(byte[] data, int offset, int length) {
        IngressStreamResult result = new IngressStreamResult();
        if (data == null || length <= 0) { result.status = Status.TRUNCATED; return result; }
        IngressKind kind = classifier.classify(data, offset, length);
        switch (kind) {
            case ENVELOPE_WIRE -> {
                Status st = parseEnvelopeFrame(data, offset, length, result);
                if (st != Status.OK) result.status = st;
            }
            case BATCH_WIRE -> wrapBatchAsEnvelope(data, offset, length, result);
            case SESSION_WIRE -> wrapSessionAsEnvelope(data, offset, length, result);
            case FIX44, SWIFT_MT940 -> wrapTextAsEnvelope(data, offset, length, kind, result);
            default -> result.status = Status.UNKNOWN_FORMAT;
        }
        return result;
    }

    public IngressStreamResult processIngressStream(byte[] data) {
        return processIngressStream(data, 0, data != null ? data.length : 0);
    }

    public void commitIngressSweep() {
        NativeBridge.nativeCommitIngressSweep();
    }

    private Status parseEnvelopeFrame(byte[] data, int offset, int length, IngressStreamResult result) {
        if (length < ENVELOPE_HEADER_SIZE) return Status.TRUNCATED;
        WireEnvelopeHeader header = new WireEnvelopeHeader();
        header.magic = BoundsUtil.readU32Le(data, offset);
        header.version = BoundsUtil.readU16Le(data, offset + 4);
        header.channelCount = BoundsUtil.readU16Le(data, offset + 6);
        header.flags = BoundsUtil.readU32Le(data, offset + 8);
        header.ingressSeq = BoundsUtil.readU32Le(data, offset + 12);
        header.parentBatchId = BoundsUtil.readU32Le(data, offset + 16);
        header.sealDigest = BoundsUtil.readU32Le(data, offset + 20);
        if (header.magic != WireTypes.ENVELOPE_MAGIC) return Status.INVALID_MAGIC;
        if (header.channelCount > config.maxChannels) return Status.BOUNDS_ERROR;
        int channelBase = offset + ENVELOPE_HEADER_SIZE;
        long need = (long) header.channelCount * ENVELOPE_CHANNEL_SIZE;
        if (!BoundsUtil.sectionBodyInBounds(channelBase, need, offset + length)) return Status.BOUNDS_ERROR;
        List<WireEnvelopeChannel> channels = new ArrayList<>();
        int payloadStart = channelBase + (int) need;
        byte[] ownedPayload = Arrays.copyOfRange(data, payloadStart, offset + length);
        for (int i = 0; i < header.channelCount; i++) {
            int base = channelBase + i * ENVELOPE_CHANNEL_SIZE;
            WireEnvelopeChannel ch = new WireEnvelopeChannel();
            ch.channelId = BoundsUtil.readU32Le(data, base);
            ch.kind = BoundsUtil.readU32Le(data, base + 4);
            ch.payloadOffset = BoundsUtil.readU32Le(data, base + 8);
            ch.payloadLen = BoundsUtil.readU32Le(data, base + 12);
            ch.routeHint = BoundsUtil.readU32Le(data, base + 16);
            channels.add(ch);
        }
        if (config.validateDigests && header.sealDigest != 0) {
            if (DigestUtil.fnv1a32(ownedPayload) != header.sealDigest) return Status.BOUNDS_ERROR;
        }
        IngressEnvelope env = buildEnvelope(header, channels, ownedPayload);
        result.completedEnvelopes.add(env);
        result.framesParsed++;
        return Status.OK;
    }

    private IngressEnvelope buildEnvelope(WireEnvelopeHeader header, List<WireEnvelopeChannel> channels, byte[] payload) {
        IngressEnvelope env = new IngressEnvelope();
        env.wire.header = header;
        env.wire.channels = channels;
        env.ownedPayload = payload;
        env.ingressSeq = header.ingressSeq != 0 ? header.ingressSeq : nextIngressSeq++;
        env.sourceKind = IngressKind.ENVELOPE_WIRE;
        for (WireEnvelopeChannel ch : channels) {
            ChannelView view = new ChannelView();
            view.channelId = ch.channelId;
            view.kind = ch.kind;
            view.routeHint = ch.routeHint;
            view.payloadLen = ch.payloadLen;
            if (ch.payloadLen > 0 && BoundsUtil.sliceInBounds(ch.payloadOffset, ch.payloadLen, payload.length)) {
                view.payload = Arrays.copyOfRange(payload, ch.payloadOffset, ch.payloadOffset + ch.payloadLen);
            } else {
                view.payload = new byte[0];
            }
            view.queued = config.queueDeferred && (header.flags & WireTypes.ENVELOPE_FLAG_SEAL_PENDING) != 0;
            env.channelViews.add(view);
            if (view.queued && view.payloadLen > 0) {
                NativeBridge.nativeQueueChannelEntry(0, view.payloadLen);
            }
        }
        return env;
    }

    private void wrapBatchAsEnvelope(byte[] data, int offset, int length, IngressStreamResult result) {
        BatchWireCodec codec = new BatchWireCodec();
        BatchWireCodec.BatchDecodeResult decoded = codec.decodeBatch(data, offset, length);
        if (decoded.status != Status.OK) { result.status = decoded.status; return; }
        WireEnvelopeHeader header = new WireEnvelopeHeader();
        header.magic = WireTypes.ENVELOPE_MAGIC;
        header.version = WireTypes.WIRE_VERSION;
        header.channelCount = 1;
        header.flags = WireTypes.ENVELOPE_FLAG_NESTED_BATCH;
        header.parentBatchId = decoded.frame.header.deskId;
        header.sealDigest = decoded.frame.header.payloadDigest;
        WireEnvelopeChannel ch = new WireEnvelopeChannel();
        ch.channelId = 1;
        ch.kind = IngressKind.BATCH_WIRE.ordinal();
        ch.payloadLen = length;
        result.completedEnvelopes.add(buildEnvelope(header, List.of(ch), Arrays.copyOfRange(data, offset, offset + length)));
        result.framesParsed++;
    }

    private void wrapSessionAsEnvelope(byte[] data, int offset, int length, IngressStreamResult result) {
        SessionWireCodec codec = new SessionWireCodec();
        SessionWireCodec.SessionDecodeResult decoded = codec.decodeSession(data, offset, length);
        if (decoded.status != Status.OK) { result.status = decoded.status; return; }
        WireEnvelopeHeader header = new WireEnvelopeHeader();
        header.magic = WireTypes.ENVELOPE_MAGIC;
        header.version = WireTypes.WIRE_VERSION;
        header.channelCount = 1;
        header.flags = WireTypes.ENVELOPE_FLAG_INGRESS_SWEEP;
        header.sealDigest = decoded.frame.header.mergeDigest;
        WireEnvelopeChannel ch = new WireEnvelopeChannel();
        ch.channelId = decoded.frame.header.sessionId;
        ch.kind = IngressKind.SESSION_WIRE.ordinal();
        ch.payloadLen = length;
        result.completedEnvelopes.add(buildEnvelope(header, List.of(ch), Arrays.copyOfRange(data, offset, offset + length)));
        result.framesParsed++;
    }

    private void wrapTextAsEnvelope(byte[] data, int offset, int length, IngressKind kind, IngressStreamResult result) {
        WireEnvelopeHeader header = new WireEnvelopeHeader();
        header.magic = WireTypes.ENVELOPE_MAGIC;
        header.version = WireTypes.WIRE_VERSION;
        header.channelCount = 1;
        header.flags = WireTypes.ENVELOPE_FLAG_CHANNEL_TAPE;
        header.sealDigest = DigestUtil.fnv1a32(data, offset, length);
        WireEnvelopeChannel ch = new WireEnvelopeChannel();
        ch.channelId = nextIngressSeq;
        ch.kind = kind.ordinal();
        ch.payloadLen = length;
        result.completedEnvelopes.add(buildEnvelope(header, List.of(ch), Arrays.copyOfRange(data, offset, offset + length)));
        result.framesParsed++;
    }
// --- expanded helpers ---
    public int envelopeCount(IngressStreamResult result) {
        return result.completedEnvelopes.size();
    }

    public void reset() {
        nextIngressSeq = 0;
    }

    public Status attachChannelViews(IngressEnvelope envelope) {
        if (envelope == null) return Status.BOUNDS_ERROR;
        envelope.channelViews.clear();
        for (WireEnvelopeChannel ch : envelope.wire.channels) {
            if (ch.payloadLen == 0) continue;
            if (!BoundsUtil.sliceInBounds(ch.payloadOffset, ch.payloadLen, envelope.ownedPayload.length)) {
                return Status.BOUNDS_ERROR;
            }
            ChannelView view = new ChannelView();
            view.channelId = ch.channelId;
            view.kind = ch.kind;
            view.payloadLen = ch.payloadLen;
            view.routeHint = ch.routeHint;
            view.payload = Arrays.copyOfRange(envelope.ownedPayload, ch.payloadOffset, ch.payloadOffset + ch.payloadLen);
            view.queued = false;
            envelope.channelViews.add(view);
        }
        return Status.OK;
    }

    public Status queueEnvelopeChannelViews(IngressEnvelope envelope) {
        if (envelope == null) return Status.BOUNDS_ERROR;
        if (!config.queueDeferred) return Status.OK;
        for (ChannelView view : envelope.channelViews) view.queued = true;
        return Status.OK;
    }

    public Status sealPendingEnvelopes(List<IngressEnvelope> pending) {
        for (IngressEnvelope envelope : pending) {
            int seal = DigestUtil.FNV_OFFSET;
            for (ChannelView view : envelope.channelViews) {
                if (view.payload != null && view.payloadLen > 0) {
                    seal = DigestUtil.mixDigest(DigestUtil.fnv1a32(view.payload, 0, view.payloadLen), seal);
                }
            }
            envelope.wire.header.sealDigest = seal;
        }
        return Status.OK;
    }

    public IngressStreamResult processMixedStream(byte[] data, int offset, int length) {
        IngressStreamResult result = new IngressStreamResult();
        String view = new String(data, offset, length, java.nio.charset.StandardCharsets.US_ASCII);
        Status fixSt = dispatchFixSegment(view, result);
        if (fixSt != Status.OK) {
            Status swiftSt = dispatchSwiftSegment(view, result);
            if (swiftSt != Status.OK) {
                result.status = dispatchBatchSegment(data, offset, length, result);
            }
        }
        return result;
    }

    private Status dispatchFixSegment(String segment, IngressStreamResult out) {
        if (segment == null || !segment.startsWith("8=FIX")) return Status.UNKNOWN_FORMAT;
        out.framesParsed++;
        IngressEnvelope envelope = new IngressEnvelope();
        envelope.sourceKind = IngressKind.FIX44;
        envelope.ingressSeq = ++nextIngressSeq;
        envelope.wire.header.magic = WireTypes.ENVELOPE_MAGIC;
        envelope.wire.header.version = WireTypes.WIRE_VERSION;
        envelope.wire.header.flags = WireTypes.ENVELOPE_FLAG_INGRESS_SWEEP;
        envelope.wire.header.ingressSeq = envelope.ingressSeq;
        envelope.ownedPayload = segment.getBytes(java.nio.charset.StandardCharsets.US_ASCII);
        WireEnvelopeChannel ch = new WireEnvelopeChannel();
        ch.channelId = envelope.ingressSeq;
        ch.kind = IngressKind.FIX44.ordinal();
        ch.payloadOffset = 0;
        ch.payloadLen = envelope.ownedPayload.length;
        envelope.wire.channels.add(ch);
        Status attach = attachChannelViews(envelope);
        if (attach != Status.OK) return attach;
        queueEnvelopeChannelViews(envelope);
        out.completedEnvelopes.add(envelope);
        return Status.OK;
    }

    private Status dispatchSwiftSegment(String segment, IngressStreamResult out) {
        if (segment == null || !segment.contains(":20:")) return Status.UNKNOWN_FORMAT;
        out.framesParsed++;
        IngressEnvelope envelope = new IngressEnvelope();
        envelope.sourceKind = IngressKind.SWIFT_MT940;
        envelope.ingressSeq = ++nextIngressSeq;
        envelope.wire.header.magic = WireTypes.ENVELOPE_MAGIC;
        envelope.wire.header.version = WireTypes.WIRE_VERSION;
        envelope.wire.header.flags = WireTypes.ENVELOPE_FLAG_CHANNEL_TAPE;
        envelope.wire.header.ingressSeq = envelope.ingressSeq;
        envelope.ownedPayload = segment.getBytes(java.nio.charset.StandardCharsets.US_ASCII);
        WireEnvelopeChannel ch = new WireEnvelopeChannel();
        ch.channelId = envelope.ingressSeq;
        ch.kind = IngressKind.SWIFT_MT940.ordinal();
        ch.payloadOffset = 0;
        ch.payloadLen = envelope.ownedPayload.length;
        envelope.wire.channels.add(ch);
        Status attach = attachChannelViews(envelope);
        if (attach != Status.OK) return attach;
        queueEnvelopeChannelViews(envelope);
        out.completedEnvelopes.add(envelope);
        return Status.OK;
    }

    private Status dispatchBatchSegment(byte[] data, int offset, int length, IngressStreamResult out) {
        BatchWireCodec codec = new BatchWireCodec();
        BatchWireCodec.BatchDecodeResult decoded = codec.decodeBatch(data, offset, length);
        if (decoded.status != Status.OK) return decoded.status;
        out.framesParsed++;
        wrapBatchAsEnvelope(data, offset, length, out);
        return Status.OK;
    }

    public static final class IngressStreamStats {
        public int framesSeen;
        public int fixMessages;
        public int swiftStatements;
        public int batchesDecoded;
        public int envelopesQueued;
        public int rejectedFrames;
    }

    public IngressStreamStats statsFor(IngressStreamResult result) {
        IngressStreamStats stats = new IngressStreamStats();
        stats.framesSeen = result.framesParsed;
        stats.envelopesQueued = result.completedEnvelopes.size();
        return stats;
    }

    public Status buildEnvelopeFromBatch(BatchWireCodec.BatchDecodeResult batchResult, IngressEnvelope envelope) {
        if (envelope == null || batchResult == null) return Status.BOUNDS_ERROR;
        envelope.wire.header.magic = WireTypes.ENVELOPE_MAGIC;
        envelope.wire.header.version = WireTypes.WIRE_VERSION;
        envelope.wire.header.channelCount = 1;
        envelope.wire.header.flags = WireTypes.ENVELOPE_FLAG_NESTED_BATCH;
        envelope.wire.header.ingressSeq = ++nextIngressSeq;
        envelope.wire.header.parentBatchId = batchResult.frame.header.deskId;
        envelope.wire.header.sealDigest = 0;
        WireEnvelopeChannel ch = new WireEnvelopeChannel();
        ch.channelId = 1;
        ch.kind = IngressKind.BATCH_WIRE.ordinal();
        ch.payloadOffset = 0;
        ch.payloadLen = batchResult.frame.payloadBlob.length;
        ch.routeHint = batchResult.frame.header.flags;
        envelope.wire.channels = List.of(ch);
        envelope.ownedPayload = batchResult.frame.payloadBlob;
        envelope.sourceKind = IngressKind.BATCH_WIRE;
        envelope.ingressSeq = envelope.wire.header.ingressSeq;
        return attachChannelViews(envelope);
    }

    public int totalChannelPayloadBytes(IngressStreamResult result) {
        int sum = 0;
        for (IngressEnvelope env : result.completedEnvelopes) {
            for (ChannelView view : env.channelViews) sum += view.payloadLen;
        }
        return sum;
    }

    public List<ChannelView> flattenChannelViews(IngressStreamResult result) {
        List<ChannelView> flat = new ArrayList<>();
        for (IngressEnvelope env : result.completedEnvelopes) flat.addAll(env.channelViews);
        return flat;
    }

    public Status validateEnvelopeBounds(IngressEnvelope envelope) {
        if (envelope == null) return Status.BOUNDS_ERROR;
        for (WireEnvelopeChannel ch : envelope.wire.channels) {
            if (ch.payloadLen > 0 && !BoundsUtil.sliceInBounds(ch.payloadOffset, ch.payloadLen, envelope.ownedPayload.length)) {
                return Status.BOUNDS_ERROR;
            }
        }
        return Status.OK;
    }

    public IngressEnvelope cloneEnvelope(IngressEnvelope source) {
        IngressEnvelope copy = new IngressEnvelope();
        copy.wire.header = source.wire.header;
        copy.wire.channels = new ArrayList<>(source.wire.channels);
        copy.ownedPayload = Arrays.copyOf(source.ownedPayload, source.ownedPayload.length);
        copy.sourceKind = source.sourceKind;
        copy.ingressSeq = source.ingressSeq;
        attachChannelViews(copy);
        return copy;
    }

    public String summarizeStream(IngressStreamResult result) {
        return "frames=" + result.framesParsed + " envelopes=" + result.completedEnvelopes.size()
                + " status=" + result.status;
    }
}

File: legacy-cpp/src/ingress/ingress_dispatch.cc
#include "tkr/ingress/ingress_dispatch.h"

#include "tkr/util/bounds.h"

#include <cstring>

namespace tkr {
namespace ingress {
namespace {

std::uint32_t ReadU32Le(const std::uint8_t* data) {
  return static_cast<std::uint32_t>(data[0]) |
         (static_cast<std::uint32_t>(data[1]) << 8) |
         (static_cast<std::uint32_t>(data[2]) << 16) |
         (static_cast<std::uint32_t>(data[3]) << 24);
}

std::uint16_t ReadU16Le(const std::uint8_t* data) {
  return static_cast<std::uint16_t>(data[0]) |
         (static_cast<std::uint16_t>(data[1]) << 8);
}

struct EnvelopeWireLayout {
  WireEnvelopeHeader header;
  std::vector<WireEnvelopeChannel> channels;
  std::size_t total_bytes;
};

Status DecodeEnvelopeWire(const std::uint8_t* data, std::size_t size,
                          EnvelopeWireLayout* out) {
  if (out == nullptr || data == nullptr || size < 24) {
    return Status::kTruncated;
  }
  out->header.magic = ReadU32Le(data + 0);
  out->header.version = ReadU16Le(data + 4);
  out->header.channel_count = ReadU16Le(data + 6);
  out->header.flags = ReadU32Le(data + 8);
  out->header.ingress_seq = ReadU32Le(data + 12);
  out->header.parent_batch_id = ReadU32Le(data + 16);
  out->header.seal_digest = ReadU32Le(data + 20);

  if (out->header.magic != kEnvelopeMagic) {
    return Status::kInvalidMagic;
  }
  if (out->header.channel_count > kMaxEnvelopeChannels) {
    return Status::kBoundsError;
  }

  const std::size_t table_bytes =
      static_cast<std::size_t>(out->header.channel_count) * 20;
  if (!util::SectionBodyInBounds(24, table_bytes, size)) {
    return Status::kBoundsError;
  }

  out->channels.clear();
  out->channels.reserve(out->header.channel_count);
  std::size_t cursor = 24;
  for (std::uint16_t i = 0; i < out->header.channel_count; ++i) {
    WireEnvelopeChannel ch{};
    ch.channel_id = ReadU32Le(data + cursor + 0);
    ch.kind = ReadU32Le(data + cursor + 4);
    ch.payload_offset = ReadU32Le(data + cursor + 8);
    ch.payload_len = ReadU32Le(data + cursor + 12);
    ch.route_hint = ReadU32Le(data + cursor + 16);
    out->channels.push_back(ch);
    cursor += 20;
  }

  out->total_bytes = size;
  return Status::kOk;
}

}  // namespace

IngressDispatch::IngressDispatch(IngressDispatchConfig config)
    : config_(config),
      classifier_(IngressClassifierConfig{true, true, 10}),
      batch_codec_(wire::BatchCodecConfig{true, true, kMaxBatchRecords}),
      ingress_seq_(0) {}

void IngressDispatch::Reset() {
  pending_envelopes_.clear();
  temp_payload_heap_.clear();
  stats_ = IngressStreamStats{};
  ingress_seq_ = 0;
  fix_parser_.ResetSession();
}

Status IngressDispatch::AttachChannelViews(IngressEnvelope* envelope) {
  if (envelope == nullptr) {
    return Status::kBoundsError;
  }
  envelope->channel_views.clear();
  for (const WireEnvelopeChannel& ch : envelope->wire.channels) {
    if (ch.payload_len == 0) {
      continue;
    }
    if (!util::SliceInBounds(ch.payload_offset, ch.payload_len,
                             envelope->owned_payload.size())) {
      return Status::kBoundsError;
    }
    ChannelView view{};
    view.channel_id = ch.channel_id;
    view.kind = ch.kind;
    view.payload_ptr = envelope->owned_payload.data() + ch.payload_offset;
    view.payload_len = ch.payload_len;
    view.route_hint = ch.route_hint;
    view.queued = false;
    envelope->channel_views.push_back(view);
  }
  return Status::kOk;
}

Status IngressDispatch::QueueEnvelopeChannelViews(IngressEnvelope* envelope) {
  if (envelope == nullptr) {
    return Status::kBoundsError;
  }
  if (!config_.queue_envelope_channels) {
    return Status::kOk;
  }
  for (ChannelView& view : envelope->channel_views) {
    view.queued = true;
  }
  return Status::kOk;
}

Status IngressDispatch::BuildEnvelopeFromWire(const wire::BatchDecodeResult& batch_result,
                                              IngressEnvelope* envelope) {
  if (envelope == nullptr) {
    return Status::kBoundsError;
  }
  envelope->wire.header.magic = kEnvelopeMagic;
  envelope->wire.header.version = kWireVersion;
  envelope->wire.header.channel_count = 1;
  envelope->wire.header.flags = kEnvelopeFlagNestedBatch;
  envelope->wire.header.ingress_seq = ++ingress_seq_;
  envelope->wire.header.parent_batch_id = batch_result.frame.header.desk_id;
  envelope->wire.header.seal_digest = 0;

  WireEnvelopeChannel ch{};
  ch.channel_id = 1;
  ch.kind = static_cast<std::uint32_t>(kIngressBatchWire);
  ch.payload_offset = 0;
  ch.payload_len = static_cast<std::uint32_t>(batch_result.frame.payload_blob.size());
  ch.route_hint = batch_result.frame.header.flags;
  envelope->wire.channels = {ch};
  envelope->owned_payload = batch_result.frame.payload_blob;
  envelope->source_kind = kIngressBatchWire;
  envelope->ingress_seq = envelope->wire.header.ingress_seq;

  return AttachChannelViews(envelope);
}

Status IngressDispatch::DispatchFixSegment(std::string_view segment,
                                           IngressStreamResult* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  if (!config_.parse_fix_sessions) {
    return Status::kOk;
  }
  wire::FixParseResult parsed = fix_parser_.ParseWithSession(segment);
  if (parsed.status != Status::kOk) {
    ++out->stats.rejected_frames;
    return parsed.status;
  }
  ++out->stats.fix_messages;
  ++out->stats.frames_seen;

  if (parsed.message.msg_type == wire::FixMsgType::kAllocationInstruction ||
      parsed.message.msg_type == wire::FixMsgType::kAllocationReport) {
    IngressEnvelope envelope;
    envelope.source_kind = kIngressFix44;
    envelope.ingress_seq = ++ingress_seq_;
    envelope.wire.header.magic = kEnvelopeMagic;
    envelope.wire.header.version = kWireVersion;
    envelope.wire.header.flags = kEnvelopeFlagIngressSweep;
    envelope.wire.header.ingress_seq = envelope.ingress_seq;
    envelope.owned_payload.assign(segment.begin(), segment.end());

    WireEnvelopeChannel ch{};
    ch.channel_id = envelope.ingress_seq;
    ch.kind = static_cast<std::uint32_t>(kIngressFix44);
    ch.payload_offset = 0;
    ch.payload_len = static_cast<std::uint32_t>(segment.size());
    envelope.wire.channels.push_back(ch);

    Status attach = AttachChannelViews(&envelope);
    if (attach != Status::kOk) {
      return attach;
    }
    QueueEnvelopeChannelViews(&envelope);
    pending_envelopes_.push_back(std::move(envelope));
    ++out->stats.envelopes_queued;
  }
  return Status::kOk;
}

Status IngressDispatch::DispatchSwiftSegment(std::string_view segment,
                                             IngressStreamResult* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  if (!config_.parse_swift_statements) {
    return Status::kOk;
  }
  wire::Mt940ScanResult scanned = swift_scanner_.Scan(segment);
  if (scanned.status != Status::kOk) {
    ++out->stats.rejected_frames;
    return scanned.status;
  }
  ++out->stats.swift_statements;
  ++out->stats.frames_seen;

  IngressEnvelope envelope;
  envelope.source_kind = kIngressSwiftMt940;
  envelope.ingress_seq = ++ingress_seq_;
  envelope.wire.header.magic = kEnvelopeMagic;
  envelope.wire.header.version = kWireVersion;
  envelope.wire.header.flags = kEnvelopeFlagChannelTape;
  envelope.wire.header.ingress_seq = envelope.ingress_seq;
  envelope.owned_payload.assign(segment.begin(), segment.end());

  WireEnvelopeChannel ch{};
  ch.channel_id = envelope.ingress_seq;
  ch.kind = static_cast<std::uint32_t>(kIngressSwiftMt940);
  ch.payload_offset = 0;
  ch.payload_len = static_cast<std::uint32_t>(segment.size());
  ch.route_hint = scanned.line_count;
  envelope.wire.channels.push_back(ch);

  Status attach = AttachChannelViews(&envelope);
  if (attach != Status::kOk) {
    return attach;
  }
  QueueEnvelopeChannelViews(&envelope);
  pending_envelopes_.push_back(std::move(envelope));
  ++out->stats.envelopes_queued;
  return Status::kOk;
}

Status IngressDispatch::DispatchBatchSegment(const std::uint8_t* data, std::size_t size,
                                             IngressStreamResult* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  wire::BatchDecodeResult decoded = batch_codec_.DecodeBatch(data, size);
  if (decoded.status != Status::kOk) {
    ++out->stats.rejected_frames;
    return decoded.status;
  }
  ++out->stats.batches_decoded;
  ++out->stats.frames_seen;
  out->decoded_batches.push_back(decoded.frame);

  IngressEnvelope envelope;
  Status built = BuildEnvelopeFromWire(decoded, &envelope);
  if (built != Status::kOk) {
    return built;
  }
  QueueEnvelopeChannelViews(&envelope);
  pending_envelopes_.push_back(std::move(envelope));
  ++out->stats.envelopes_queued;
  return Status::kOk;
}

Status IngressDispatch::DispatchEnvelopeSegment(const std::uint8_t* data, std::size_t size,
                                                IngressStreamResult* out) {
  if (out == nullptr) {
    return Status::kBoundsError;
  }
  EnvelopeWireLayout layout;
  Status decoded = DecodeEnvelopeWire(data, size, &layout);
  if (decoded != Status::kOk) {
    ++out->stats.rejected_frames;
    return decoded;
  }

  std::size_t payload_offset = 24 + layout.channels.size() * 20;
  std::size_t payload_span = 0;
  for (const WireEnvelopeChannel& ch : layout.channels) {
    if (ch.payload_len == 0) {
      continue;
    }
    const std::size_t end =
        static_cast<std::size_t>(ch.payload_offset + ch.payload_len);
    if (end > payload_span) {
      payload_span = end;
    }
  }
  if (!util::SectionBodyInBounds(payload_offset, payload_span, size)) {
    return Status::kBoundsError;
  }

  auto owned = std::make_unique<std::vector<std::uint8_t>>();
  if (payload_span > 0) {
    owned->assign(data + payload_offset, data + payload_offset + payload_span);
  }

  IngressEnvelope envelope;
  envelope.wire.header = layout.header;
  envelope.wire.channels = layout.channels;
  envelope.source_kind = kIngressEnvelopeWire;
  envelope.ingress_seq = layout.header.ingress_seq;

  temp_payload_heap_.push_back(std::move(owned));
  std::vector<std::uint8_t>* heap_buf = temp_payload_heap_.back().get();

  const bool cross_frame_tape =
      (layout.header.flags & kEnvelopeFlagCrossFrameTape) != 0;
  if ((layout.header.flags & kEnvelopeFlagChannelTape) != 0 &&
      cross_frame_tape && heap_buf != nullptr) {
    envelope.owned_payload.clear();
    envelope.channel_views.clear();
    for (const WireEnvelopeChannel& ch : envelope.wire.channels) {
      if (ch.payload_len == 0) {
        continue;
      }
      if (!util::SliceInBounds(ch.payload_offset, ch.payload_len, heap_buf->size())) {
        return Status::kBoundsError;
      }
      ChannelView view{};
      view.channel_id = ch.channel_id;
      view.kind = ch.kind;
      view.payload_ptr = heap_buf->data() + ch.payload_offset;
      view.payload_len = ch.payload_len;
      view.route_hint = ch.route_hint;
      view.queued = false;
      envelope.channel_views.push_back(view);
    }
  } else {
    envelope.owned_payload = *heap_buf;
    Status attach = AttachChannelViews(&envelope);
    if (attach != Status::kOk) {
      temp_payload_heap_.pop_back();
      return attach;
    }
  }
  QueueEnvelopeChannelViews(&envelope);
  pending_envelopes_.push_back(std::move(envelope));
  ++out->stats.frames_seen;
  ++out->stats.envelopes_queued;
  return Status::kOk;
}

Status IngressDispatch::DispatchSessionSegment(const std::uint8_t* data, std::size_t size,
                                               IngressStreamResult* out) {
  if (out == nullptr || size < 24) {
    return Status::kTruncated;
  }
  const std::uint32_t magic = ReadU32Le(data);
  if (magic != kSessionMagic) {
    return Status::kInvalidMagic;
  }
  ++out->stats.frames_seen;
  return Status::kOk;
}

Status IngressDispatch::CommitIngressSweep() {
  temp_payload_heap_.clear();
  pending_envelopes_.clear();
  return Status::kOk;
}

Status IngressDispatch::SealPendingEnvelopes() {
  while (!pending_envelopes_.empty()) {
    IngressEnvelope envelope = std::move(pending_envelopes_.front());
    pending_envelopes_.pop_front();

    std::uint32_t seal = 2166136261u;
    for (const ChannelView& view : envelope.channel_views) {
      if (view.payload_ptr != nullptr && view.payload_len > 0) {
        seal = util::Fnv1a32(view.payload_ptr, view.payload_len) ^ (seal * 16777619u);
      }
    }
    envelope.wire.header.seal_digest = seal;
  }
  return Status::kOk;
}

IngressStreamResult IngressDispatch::ProcessIngressStream(const std::uint8_t* data,
                                                          std::size_t size) {
  IngressStreamResult result{};
  result.status = Status::kOk;

  if (data == nullptr || size == 0) {
    result.status = Status::kTruncated;
    return result;
  }

  IngressClassifyResult classified = classifier_.ClassifyBinary(data, size);
  if (classified.status != Status::kOk &&
      classified.primary_kind == kIngressUnknown) {
    result.status = classified.status;
    return result;
  }

  std::string_view view(reinterpret_cast<const char*>(data), size);

  switch (classified.primary_kind) {
    case kIngressFix44:
      result.status = DispatchFixSegment(view, &result);
      break;
    case kIngressSwiftMt940:
      result.status = DispatchSwiftSegment(view, &result);
      break;
    case kIngressBatchWire:
      result.status = DispatchBatchSegment(data, size, &result);
      break;
    case kIngressEnvelopeWire:
      result.status = DispatchEnvelopeSegment(data, size, &result);
      break;
    case kIngressSessionWire:
      result.status = DispatchSessionSegment(data, size, &result);
      break;
    case kIngressMixedStream: {
      Status fix_st = DispatchFixSegment(view, &result);
      if (fix_st != Status::kOk) {
        Status swift_st = DispatchSwiftSegment(view, &result);
        if (swift_st != Status::kOk) {
          result.status = DispatchBatchSegment(data, size, &result);
        }
      }
      break;
    }
    default:
      result.status = Status::kUnknownFormat;
      ++result.stats.rejected_frames;
      return result;
  }

  if (result.status == Status::kOk && !pending_envelopes_.empty()) {
    for (IngressEnvelope& env : pending_envelopes_) {
      result.completed_envelopes.push_back(env);
    }
  }

  result.stats = stats_;
  return result;
}

IngressStreamResult IngressDispatch::ProcessIngressStream(std::string_view text) {
  return ProcessIngressStream(reinterpret_cast<const std::uint8_t*>(text.data()),
                              text.size());
}

}  // namespace ingress
}  // namespace tkr
